node/runtime/pangolin/src/pallets/from_substrate_issuing_.rs

```diff
@@ -5,55 +5,37 @@ use sp_runtime::AccountId32;
 // --- darwinia-network ---
 use crate::*;
 use bridge_primitives::{AccountIdConverter, PANGORO_CHAIN_ID};
-use darwinia_support::{s2s::ToEthAddress, to_bytes32, ChainName};
-use dp_asset::token::Token;
-use dp_contract::mapping_token_factory::s2s::S2sRemoteUnlockInfo;
-use from_substrate_issuing::{Config, EncodeCall};
+use darwinia_support::{s2s::ToEthAddress, ChainName};
+use dp_s2s::{CallParams, PayloadCreate};
+use from_substrate_issuing::Config;
 
-// remote chain pangoro's dispatch info
-#[derive(Encode, Decode, Debug, PartialEq, Eq, Clone)]
-pub enum PangoroRuntime {
-	/// s2s bridge backing pallet.
-	/// this index must be the same as the backing pallet in pangoro runtime
-	#[codec(index = 20)]
-	Sub2SubBacking(PangoroSub2SubBackingCall),
-}
-
-#[derive(Encode, Decode, Debug, PartialEq, Eq, Clone)]
-#[allow(non_camel_case_types)]
-pub enum PangoroSub2SubBackingCall {
-	#[codec(index = 2)]
-	unlock_from_remote(Token, AccountId),
-}
-
-pub struct PangoroCallEncoder;
-impl EncodeCall<AccountId, ToPangoroMessagePayload> for PangoroCallEncoder {
-	fn encode_remote_unlock(
-		submitter: AccountId,
-		remote_unlock_info: S2sRemoteUnlockInfo,
-	) -> Result<ToPangoroMessagePayload, ()> {
-		if remote_unlock_info.recipient.len() != 32 {
-			return Err(());
-		} else {
-			let recipient_id: AccountId =
-				to_bytes32(remote_unlock_info.recipient.as_slice()).into();
-			let call =
-				PangoroRuntime::Sub2SubBacking(PangoroSub2SubBackingCall::unlock_from_remote(
-					remote_unlock_info.token,
-					recipient_id,
-				))
-				.encode();
-			return Ok(ToPangoroMessagePayload {
-				spec_version: remote_unlock_info.spec_version,
-				weight: remote_unlock_info.weight,
-				origin: bp_message_dispatch::CallOrigin::SourceAccount(submitter),
-				call,
-				dispatch_fee_payment: DispatchFeePayment::AtSourceChain,
-			});
-		}
+const PANGORO_S2S_BACKING_PALLET_INDEX: u8 = 20;
+/// Create message payload according to the call parameters.
+pub struct PangoroPayloadCreator;
+impl PayloadCreate<AccountId, ToPangoroMessagePayload> for PangoroPayloadCreator {
+	fn payload(
+		spec_version: u32,
+		weight: u64,
+		call_params: CallParams<AccountId>,
+	) -> Result<ToPangoroMessagePayload, &'static str> {
+		let (submitter, call) = match call_params.clone() {
```

Comment:
still, I think this match is complex. and the type CallParams is not intuitive.

Reply:
After s2s-audit is merged, the payload() signature has a submitter param, this match can be deleted.